cineast-core/src/main/java/org/vitrivr/cineast/core/temporal/timedistance/TimeDistanceTemporalScoringAlgorithm.java

```diff
@@ -0,0 +1,162 @@
+      SortedSet<ScoredSegment> setFromElement = this.scoredSegmentSets.get(currentSegment.getObjectId()).tailSet(item);
+      List<ScoredSegment> candidates = setFromElement.stream().filter(c -> c.getContainerId() == lambdaInnerId).filter(c -> c.getEndAbs() - item.getStartAbs() >= 0).filter(c -> c.getEndAbs() - item.getStartAbs() <= this.maxLength).filter(c -> !c.getSegmentId().equals(item.getSegmentId())).collect(Collectors.toList());
+      if (candidates.size() == 0) {
+        continue;
+      }
+      ScoredSegment bestSegment = null;
+      double bestScore = 0D;
+
+      /* Go through all candidates from the currentSegment */
+      for (ScoredSegment candidate : candidates) {
+        MediaSegmentDescriptor innerDescriptor = this.segmentMap.get(candidate.getSegmentId());
+        if (innerDescriptor == null) {
+          continue;
+        }
+        /* Calculate the inverse decay score of the candidate */
+        double innerScore = calculateInverseDecayScore(currentEndAbs, candidate, this.timeDistances.get(innerContainerId - 1), innerDescriptor);
+        if (innerScore > bestScore) {
+          bestScore = innerScore;
+          bestSegment = candidate;
+        }
+      }
+      /* If there is a best segment, store it and update current End abs, otherwise update end abs with time distances */
+      if (bestSegment != null) {
+        currentSegment = bestSegment;
+        currentEndAbs = this.segmentMap.get(bestSegment.getSegmentId()).getEndabs();
+        segments.add(bestSegment.getSegmentId());
+      } else {
+        currentEndAbs += timeDistances.get(innerContainerId - 1);
+      }
+      score += bestScore;
+    }
+
+    return new TemporalObject(segments, item.getObjectId(), score / (this.maxContainerId + 1));
+  }
+
+  private double calculateInverseDecayScore(float currentSegmentEndTime, ScoredSegment nextSegment, float timeDifference, MediaSegmentDescriptor segmentDescriptor) {
+    double score = 0D;
+    if (segmentDescriptor.getStartabs() >= currentSegmentEndTime && segmentDescriptor.getStartabs() < currentSegmentEndTime + timeDifference) {
+      score = Math.exp((0.1f * (segmentDescriptor.getStartabs() - currentSegmentEndTime - timeDifference))) * nextSegment.getScore();
+    } else if (segmentDescriptor.getStartabs() >= currentSegmentEndTime && segmentDescriptor.getStartabs() > currentSegmentEndTime + timeDifference) {
+      score = Math.exp((-0.1f * (segmentDescriptor.getStartabs() - currentSegmentEndTime - timeDifference))) * nextSegment.getScore();
+    } else if (segmentDescriptor.getStartabs() == currentSegmentEndTime + timeDifference) {
+      score = nextSegment.getScore();
+    }
+    return score;
+  }
+
+  /* Storage class for the results for easier result transformation. */
+  private static class ResultStorage {
+
+    private Set<String> segments = new TreeSet<>();
```

Comment:
Is there a reason this is a TreeSet and not a Simple HashSet?

Reply:
Laziness when retrieving the results (so no good reason) but I updated it because it makes computationally probably more sense.